Cleanups before Release

JesusMcCloud opened the pull request:
Proper version of #82 going into development.  
Once merged, will prepare a release.

Changes the following:

 * Rename `toCryptoAlgorithm` to `toX509SignatureAlgorithm` accordingly
 * Rework CryptoSignature to two-dimensional interface:
   * CryptoSignature <- {EC <- {IndefiniteLength, DefiniteLength}, RsaOrHmac}
    * CryptoSignature <- {Defined <- {Ill <- EC.IndefiniteLength, Well <- {EC.DefiniteLength, RsaOrHmac}}

This finally makes it possible to get a common ground to encode RSA/HMAC and EC signatures for COSE and JWS, without having to jump through hoops. The effect of this can be seen in [the corresponding  VC Lib PR](https://github.com/a-sit-plus/kmm-vc-library/pull/83).

iaik-jheher commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt:
```diff
@@ -6,54 +6,59 @@ import at.asitplus.crypto.datatypes.asn1.DERTags.DER_SEQUENCE
-        if (other == null || this::class != other::class) return false
-
-        other as CryptoSignature
+    sealed interface Defined : CryptoSignature {
```
what's the benefit of this inheritance structure? why not just `IllDefined: CryptoSignature` and `WellDefined: CryptoSignature`? (Or even drop `IllDefined` altogether since it seems to add nothing?)

JesusMcCloud replied:
Without two mutually exlcusive Types, you won't win anything, so it needs to stay, because otherwise

this
```
//it is CryptoSignature.Defined
when (it) {
    is CryptoSignature.Defined.Well -> it
    is CryptoSignature.Defined.Ill -> when (it) {
        is CryptoSignature.EC.IndefiniteLength -> it.withCurve((publicKey as CryptoPublicKey.EC).curve)
    }
}
```

turns into this useless mess:
```
//it is CryptoSignature
when (it) {
    is CryptoSignature.Definite -> it
    else -> when (it) {
        is CryptoSignature.EC.IndefiniteLength -> it.withCurve((publicKey as CryptoPublicKey.EC).curve)
        is CryptoSignature.EC.DefiniteLength -> TODO("utterly useless")
        is CryptoSignature.RSAorHMAC -> TODO("utterly useless")
    }
}
```

iaik-jheher replied:
```kotlin
val defined = when (it) {
        is CryptoSignature.WellDefined -> it
        is CryptoSignature.EC.IndefiniteLength -> it.withCurve()
    }
```
https://pl.kotl.in/MyyqCQvUs

JesusMcCloud replied:
to me, this is not a technical issue, but one of how expressive the code is to a human. without the inverse interface, something the grouping (first code block I posted) won't be possible: https://pl.kotl.in/AX88suTB4

iaik-jheher commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/ECPoint.kt:
```diff
@@ -87,7 +90,11 @@ sealed class ECPoint private constructor(
+                homZ.toString(
+                    16
+                )
+            }]"
```
this indentation change feels particularly excessive (and quite a few of the others make the code less readable imo, too)

iaik-jheher commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/misc/PointCompression.kt:
```diff
@@ -11,29 +11,43 @@ enum class ANSIECPrefix(val prefixByte: Byte) {
+
+        /**
+         * Gets the ANSI prefix for [byte].
+         * @throws IllegalArgumentException for ZERO
```
copy/paste error?

iaik-jheher commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/misc/PointCompression.kt:
```diff
@@ -68,9 +82,9 @@ internal fun decompressY(curve: ECCurve, x: ModularBigInteger, sign: Sign): Modu
 
     require(quadraticResidueTest(alpha))
-        { "Invalid compressed point (x=$x) on $curve"}
+    { "Invalid compressed point (x=$x) on $curve" }
```
i really dislike removing the indentation here

iaik-jheher commented on CHANGELOG.md:
```diff
@@ -139,17 +139,26 @@
- * Rename `CryptoAlgorithm` to `X509SignatureAlgorithm` to better describe what it is
+
+**Changes**
+* Drop support for did-encoding uncompressed keys (but keep decoding support)
```
```suggestion
* Always DID-encode keys in compressed form (but keep decoding support)
```

nodh commented on datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt:
```diff
@@ -89,13 +87,13 @@ sealed class CoseKeyParams {
         override fun toCryptoPublicKey(): KmmResult<CryptoPublicKey> {
-            return runCatching {
-                CryptoPublicKey.EC(
+            return catching {
```
Probably makes sense to convert from block body to expression

nodh replied:
You did not address this issue imho

nodh commented on datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt:
```diff
@@ -167,13 +164,13 @@ sealed class CoseKeyParams {
 
         override fun toCryptoPublicKey(): KmmResult<CryptoPublicKey> {
-            return runCatching {
+            return catching {
```
Probably makes sense to convert from block body to expression

nodh commented on datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseSigned.kt:
```diff
@@ -87,14 +82,14 @@ data class CoseSigned(
 
 fun CoseHeader.usesEC(): Boolean? {
-    algorithm?.toCryptoAlgorithm()?.isEc?.let { return it }
+    algorithm?.toX509SignatureAlgorithm()?.isEc?.let { return it }
```
Probably can be refactored to an expression body

nodh commented on datatypes-jws/src/commonMain/kotlin/at/asitplus/crypto/datatypes/jws/JsonWebKey.kt:
```diff
@@ -281,10 +282,10 @@ data class JsonWebKey(
      */
     fun toCryptoPublicKey(): KmmResult<CryptoPublicKey> =
-        runCatching {
+        catching {
```
This `catching` can probably moved up one line, to reduce indendation in the body

nodh commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoPublicKey.kt:
```diff
@@ -347,24 +351,34 @@ sealed class CryptoPublicKey : Asn1Encodable<Asn1Sequence>, Identifiable {
-                "fromUncompressed(curve, x, y)"))
+            @Deprecated(
+                "Explicitly specify what you want", ReplaceWith(
+                    "fromUncompressed(curve, x, y)"
```
Why is this wrapped here?

JesusMcCloud replied:
let me rephrase that for you: what's the deal with the auto-formatter?

nodh commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt:
```diff
@@ -6,54 +6,59 @@ import at.asitplus.crypto.datatypes.asn1.DERTags.DER_SEQUENCE
     override fun encodeToTlv(): Asn1Element = signature
 
-    override fun toString(): String {
+    fun str(): String {
```
Method name seems a bit short, and can probably be converted into an expression

nodh commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/X509SignatureAlgorithm.kt:
```diff
@@ -11,18 +12,9 @@ import kotlinx.serialization.descriptors.SerialDescriptor
 @Serializable(with = X509SignatureAlgorithmSerializer::class)
-enum class X509SignatureAlgorithm(override val oid: ObjectIdentifier, val isEc: Boolean = false)
-    : Asn1Encodable<Asn1Sequence>, Identifiable {
+enum class X509SignatureAlgorithm(override val oid: ObjectIdentifier, val isEc: Boolean = false) :
```
Maybe wrap this the other way round?

iaik-jheher reviewed (changes requested):
.

nodh commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/io/Encoding.kt:
```diff
@@ -50,7 +51,7 @@ object ByteArrayBase64Serializer : KSerializer<ByteArray> {
      */
     override fun deserialize(decoder: Decoder): ByteArray {
-        return kotlin.runCatching { decoder.decodeString().decodeToByteArray(Base64Strict) }
+        return catching { decoder.decodeString().decodeToByteArray(Base64Strict) }
```
can be converted to an expression

nodh commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/io/Encoding.kt:
```diff
@@ -74,7 +75,7 @@ object ByteArrayBase64UrlSerializer : KSerializer<ByteArray> {
      */
     override fun deserialize(decoder: Decoder): ByteArray {
-        return kotlin.runCatching { decoder.decodeString().decodeToByteArray(Base64UrlStrict) }
+        return catching { decoder.decodeString().decodeToByteArray(Base64UrlStrict) }
```
can be converted to an expression

nodh commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/misc/PointCompression.kt:
```diff
@@ -68,9 +82,9 @@ internal fun decompressY(curve: ECCurve, x: ModularBigInteger, sign: Sign): Modu
+    { "Invalid compressed point (x=$x) on $curve" }
     require(curve.modulus.bitAt(0) && curve.modulus.bitAt(1)) // (modulus % 4) == 3
-        { "Decompression on $curve requires Tonelli-Shanks Algorithm" }
+    { "Decompression on $curve requires Tonelli-Shanks Algorithm" }
```
Same with the indendation here

nodh commented on datatypes/src/jvmMain/kotlin/at/asitplus/crypto/datatypes/JcaExtensions.kt:
```diff
@@ -80,43 +80,43 @@ fun CryptoPublicKey.getJcaPublicKey() = when (this) {
 
 fun CryptoPublicKey.EC.getJcaPublicKey(): KmmResult<ECPublicKey> {
-    return runCatching {
+    return catching {
```
can be converted to an expression

nodh commented on datatypes/src/jvmMain/kotlin/at/asitplus/crypto/datatypes/JcaExtensions.kt:
```diff
@@ -80,43 +80,43 @@ fun CryptoPublicKey.getJcaPublicKey() = when (this) {
 
 fun CryptoPublicKey.Rsa.getJcaPublicKey(): KmmResult<RSAPublicKey> =
-    runCatching {
+    catching {
```
move up this `catching` a line to reduce indendation in the body?

nodh commented on datatypes/src/jvmTest/kotlin/at/asitplus/crypto/datatypes/X509CertificateJvmTest.kt:
```diff
@@ -82,7 +83,7 @@ class X509CertificateJvmTest : FreeSpec({
             update(tbsCertificate.encodeToTlv().derEncoded)
         }.sign()
-        val test = (CryptoSignature.decodeFromDer(signed) as CryptoSignature.EC.IndefiniteLength).withCurve(ECCurve.SECP_256_R_1)
+        val test =            CryptoSignature.decodeFromDer(signed)
```
spaces here seem off

JesusMcCloud commented:
@nodh I addressed all your comments

JesusMcCloud commented:
> > @nodh I addressed all your comments
> 
> I don't think so, at least the last commit [b97f224](https://github.com/a-sit-plus/kmp-crypto/commit/b97f224afa0536e79d5d2f192c60f6fff15c4808) doesn't address my comments.

TIL: pushing commits help them get to the remote

iaik-jheher commented on datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt:
```diff
@@ -166,21 +162,33 @@ sealed class CoseKeyParams {
+        override fun toCryptoPublicKey(): KmmResult<CryptoPublicKey> = catching {
+            CryptoPublicKey.Rsa(
+                n = n ?: throw IllegalArgumentException("Missing modulus n"),
+                e = e?.let { bytes -> Int.decodeFromDer(bytes) }
+                    ?: throw IllegalArgumentException("Missing or invalid exponent e")
```
```suggestion
                e = Int.decodeFromDer(e ?:
                    throw IllegalArgumentException("Missing or invalid exponent e"))
```

iaik-jheher commented on datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt:
```diff
@@ -137,24 +147,27 @@ sealed class CryptoSignature(
 
-                require(s.bitLength() <= scalarByteLength.toInt()*8) {
-                    "s is ${s.bitLength()} bits long, expected at most ${scalarByteLength.toInt()} bytes (${max} bits)"}
+                require(s.bitLength() <= scalarByteLength.toInt() * 8) {
```
```suggestion
                require(s.bitLength() <= max) {
```
